### PR TITLE
[ConstraintElim] Look through post-increment in header EQ/NE compare.

### PR DESCRIPTION
**Pull request by fhahn**
Extend the header-controlled induction handling to look through a post-increment `PN + C` (constant C) on the compared value

For now, the new path is limited to deriving unsigned facts.

Alive2 Proof: https://alive2.llvm.org/ce/z/DCNeSr

### CODE REVIEW

**Review by dtcxzyw — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 1]
```diff
@@ -1003,6 +1009,11 @@ void State::addInfoForInductions(BasicBlock &BB) {
   if (!L->isLoopInvariant(B))
     return;
 
+  // If we looked through `PN + C`, only derive facts when that add is
+  // really the induction's post-increment.
+  if (IncStep && (*IncStep != StepOffset || StepOffset.isNegative()))
```
Comment (antoniofrighetto) [inline llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1013]: Maybe nit: should we add a test rejecting negative steps / anticipate the TODO below?

**Review by antoniofrighetto — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp (hunk 1 above)
Comment (fhahn) [reply llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1013]: Thanks, added additional tests to cover the previously missing code paths

**Comment by fhahn** [comment]
Follow-up to derive signed facts: https://github.com/llvm/llvm-project/pull/210079
